### DISCUSSION TITLE
Built in script execution support

### DISCUSSION THREAD

**Discussion by kopach** [espanso/espanso#1410] [Feature requests and Ideas]
## Problem
For more complex operations, there is a need to do those using some programming language. 
Currently, espanso supports usage of external CLI apps for those purposes. 
The disadvantage of that, is that as plugin author, when I build a plugin which I would like to share with the community - I need also to provide instructions or prerequisites (e.g. make sure to have node.js installed or so).

## Proposal
It would be great, if espanso could execute some scripting language without a need to install an engine for that. So, I, as a plugin author, could share a plugin with a script only.

I would recommend **JavaScript** as it one of the most popular scripting language at the moment. Recent versions of Node.js can also run [WebAssembly](https://nodejs.dev/en/learn/nodejs-with-webassembly/), if performance is needed for some cases.

**Comment by kopach** [comment]
Another benefit of integrated Node.js runtime into Espanso itself is "cold start" problem. Basically, when you run a script with Node.js for the first time - it takes extremely more time than in subsequent runs. E.g. this simple script `node -p  require('./package.json').version"` took me `1.641 seconds` to run first time and only `94 ms` to run next time - **30x faster**.

Because Espanso commands might not be triggered very often, most of the runs would be slow due to cold start of language engine.

So, Espanso could keep engine "pre-warmed" for all commands

**Reply by kopach** [reply]
@federico-terzi, do you think this idea is legit and something like this could be possible to implement?

**Reply by smeech** [reply]
Sorry this has hung for so long. I'm just trying to progress some of the Issues and Discussions.

Does Espanso invoke JavaScript natively currently? I have only come across Python scripts, although I guess others can be called via the Shell extension.

**Reply by kopach** [reply]
Espanso can invoke JS as node.js script from shell. But this requires user to have `node.js` installed. I believe same is true for Python, python must be installed for scripts to work, right?

**Reply by smeech** [reply]
Ah, I thought it might have to be via the `shell `extension, rather than the `script` one. An alternative might be sharing of an executable, but that has greater security concerns. 

Yes, I can see the parallels. Perhaps we ought to be seeking an addition to `script` capabilities? Thank you!

I'll flag this as an "enhancement" for now.

**Comment by reitzig** [comment]
FWIW: I for one would be happy with a well-defined way to ship scripts in packages (#1558). I can take care of the necessary runtime environment (if not always _easily_ so 😖).

**Comment by reitzig** [comment]
Isn't Lua a more popular choice for "built-in scripting"?

**Reply by kopach** [reply]
Could be, not sure. But I think here is more important to get wider audience and adoption for developers. Lua isn't very popular as a language, unfortunately, based on today's trends.

**Reply by kpym** [reply]
Yes, but Lua is small and `nodejs` is big. Personally, I don't want to see `nodejs` end up in espanso for this reason _(and many others)_.

**Reply by kopach** [reply]
Why not C++ then? It can be small. Or Java? It runs in every SIM card in our phones.
IMO, the key here is to make writing plugins as much accessible as possible, and JS is most "known" language nowadays.
Also, there are other newer, faster & "smaller" JS runtimes nowadays like [bun](https://bun.sh/), [deno](https://deno.com/). Some can even run TypeScript natively. 
Btw, none of those engines not necessary should be shipped with Espanso by default. It could be based on features used. E.g., when user would install/use plugin/functionality which requires JS runtime - there could be question like:

> This plugin requires additional resources to be installed. Do you want to continue?
> Will be installed: x, y, z

If `yes` -> runtime can be downloaded add added as plugin to Espanso.

**Reply by kpym** [reply]
If we should add some extra stuff, why not use an external interpreter in this case? For javascript you can just put a single executable like [quickjs](https://bellard.org/quickjs) in `%CONFIG%/bin` and call it with the `script` variable. 

Maybe you just want to make the installation of this extra interpreter in `%CONFIG%/bin` automatic in some cases ?

**Reply by kopach** [reply]
Just found a good example of such additional dependency management in Alfred. It's only for macOS, so it's simpler, but could be a good starting point  https://www.alfredapp.com/help/kb/dependencies/

**Comment by smeech** [comment]
I don't think we should be packaging script interpreters with Espanso packages - there are several, and people have their own preferences. I wouldn't want to see the additional complexity of maintaining updates and addressing conflicts on users' systems.
I think the `script:` extension should have the facility connect to as wide as possible a range of options (what are its advantages over a `shell:` `cmd` call?).
It's not too much to ask to make `readme.md`s explicit, and have prompts on installation, as suggested above.

**Comment by kpym** [comment]
>  I think the `script:` extension should have the facility connect to as wide as possible a range of options (what are its advantages over a `shell:` `cmd` call?).

- `script` is lighter and faster _(no need to run a shell just to run another program)_ ;
- `script` is shell/os independant.

**Reply by smeech** [reply]
Thank you - good points.
Looking through the hub repo, I can't find any scripts used currently - not even Python. This may be partly due to the difficulties described above, I suppose.

**Comment by kopach** [comment]
Just checked what does main competitor of Espanso offers - they have JavaScript execution built in (alongside of shell & Apple Script)
https://textexpander.com/learn/using/snippets/scripting-textexpander
IMO, adding built in JS execution is logical next step in evolution of Espanso. Shell scripting has a big disadvantage - it is very OS dependent

**Reply by reitzig** [reply]
I for one don't think that a comparison of Espanso with a commercial Mac-only product should inform the roadmap of Espanso all too much. 

IMO, integrating any form of JS stack in what could easily be a keylogger is a major risk. A company can mitigate such risks (and make users pay for it) -- OSS maintainers can't, really. (In the sense of, we can't expect it of them -- and so we can't really trust the software.)

**Reply by kopach** [reply]
Well, make sense, haven't thought about this aspect. Actually, more I think about it, more I believe Espanso should have control over the runtime environment (even with a cost of extra space). Allowing anyone to ship binaries which are difficult to check & restrict is more dangerous.
As an option, JS engine like [Deno](https://deno.com/) has all accesses restricted by default (e.g. network, file system access, etc), so it's easier to control what each plugin has access to and explicitly enabling those if needed

**Comment by smeech** [comment]
In a similar vein, it would enhance Espanso if we could use a wider range of shells, e.g. PowerShell in Linux.
